base/reduce.jl

```diff
@@ -47,18 +47,29 @@ function foldl_impl(op::OP, nt, itr) where {OP}
     return v
 end
 
-function _foldl_impl(op::OP, init, itr) where {OP}
-    # Unroll the while loop once; if init is known, the call to op may
-    # be evaluated at compile time
+function _foldl_impl(op::OP, init::T, itr) where {OP, T}
     y = iterate(itr)
     y === nothing && return init
     v = op(init, y[1])
+    if v isa T
+        return _foldl_impl(op, v, itr, y[2])
+    else
+        return _foldl_impl(op, v, itr, y[2])
```

Comment:
Isn't this exactly the same as what's done in the other branch of the conditional? :thinking:

Reply:
Yes, but it helps the compiler.  Removing this branch introduces an allocation and makes the code a bit slower:

```julia
julia> @eval Base function _foldl_impl(op::OP, init::T, itr) where {OP, T}
           y = iterate(itr)
           y === nothing && return init
           v = op(init, y[1])
           return _foldl_impl(op, v, itr, y[2])
       end
_foldl_impl (generic function with 4 methods)

julia> @btime sum(x for x in $xs if x !== missing)
  780.859 ns (1 allocation: 16 bytes)
```

But the speedup is not as drastic as I felt while I was playing with it. So I'm OK with removing this micro-optimization.

Reply:
Actually, as this "no-op `if` branch" gets rid of a constant-time work, the difference becomes large if you consider shorter arrays:

```julia
julia> xs = [abs(x) < 1 ? x : missing for x in randn(10)];

julia> @btime sum(x for x in $xs if x !== missing)
  8.405 ns (0 allocations: 0 bytes)
1.5696312630017393

julia> @eval Base function _foldl_impl(op::OP, init::T, itr) where {OP, T}
           y = iterate(itr)
           y === nothing && return init
           v = op(init, y[1])
           return _foldl_impl(op, v, itr, y[2])
       end
_foldl_impl (generic function with 4 methods)

julia> @btime sum(x for x in $xs if x !== missing)
  29.120 ns (1 allocation: 16 bytes)
1.5696312630017393
```

Reply:
That is incredibly bizarre. :flushed:

Reply:
I thought I was just doing union splitting manually.